subprojects/build-option/src/main/java/org/gradle/internal/buildoption/BuildOption.java

```diff
@@ -30,19 +30,31 @@
  */
 public interface BuildOption<T> {
 
-    @Nullable String getGradleProperty();
+    @Nullable
+    String getGradleProperty();
+
     void applyFromProperty(Map<String, String> properties, T settings);
+
     void configure(CommandLineParser parser);
+
     void applyFromCommandLine(ParsedCommandLine options, T settings);
 
     enum Origin {
-        GRADLE_PROPERTY, COMMAND_LINE;
-
-        public void handleInvalidValue(String gradlePropertyFailureMessage, String commandLineFailureMessage) {
-            switch (this) {
-                case GRADLE_PROPERTY: throw new IllegalArgumentException(gradlePropertyFailureMessage);
-                case COMMAND_LINE: throw new CommandLineArgumentException(commandLineFailureMessage);
+        GRADLE_PROPERTY {
+            @Override
+            public void handleInvalidValue(String property, String option, String value, String hint) {
```

Comment:
The `hint` parameter should be optional. You might want to introduce two different methods.

Reply:
Done.